src/cpp/session/modules/SessionCopilot.cpp

```diff
@@ -0,0 +1,894 @@
+   textDocumentJson["uri"] = uriFromDocumentId(pDoc->id());
+   textDocumentJson["languageId"] = languageIdFromDocument(pDoc);
+   textDocumentJson["version"] = 1;
+   textDocumentJson["text"] = pDoc->contents();
+
+   json::Object paramsJson;
+   paramsJson["textDocument"] = textDocumentJson;
+
+   sendNotification("textDocument/didOpen", paramsJson);
+}
+
+void onDocRemoved(const std::string& id, const std::string& path)
+{
+   if (!ensureAgentRunning())
+      return;
+
+   json::Object textDocumentJson;
+   textDocumentJson["uri"] = uriFromDocumentId(id);
+
+   json::Object paramsJson;
+   paramsJson["textDocument"] = textDocumentJson;
+
+   sendNotification("textDocument/didClose", paramsJson);
+}
+
+void onBackgroundProcessing(bool isIdle)
+{
+   // extract requests that appear to have been dropped
+   if (isIdle)
+   {
+      auto currentTime = boost::posix_time::second_clock::local_time();
+
+      std::vector<std::string> keys;
+      boost::copy(
+               s_pendingContinuations | boost::adaptors::map_keys,
+               std::back_inserter(keys));
+
+      for (auto&& key : keys)
+      {
+         auto&& continuation = s_pendingContinuations[key];
+         auto elapsedTime = currentTime - continuation.time();
+         if (elapsedTime.seconds() > 10)
+         {
+            DLOG("Cancelling old contiuation with id '{}'.", key);
+            continuation.cancel();
+            s_pendingContinuations.erase(key);
+         }
+      }
+   }
+
+   // process any pending requests
+   while (!s_pendingResponses.empty())
+   {
+      // Try to parse this as JSON, and see if we received the response we expect.
+      // The agent might be chatty inbetween a request and a response, so we can
+      // drop any output not specifically for us.
+      std::string response = s_pendingResponses.front();
+      s_pendingResponses.pop();
+
+      json::Object responseJson;
+      Error error = responseJson.parse(response);
+      if (error)
+      {
+         LOG_ERROR(error);
+         continue;
+      }
+
+      // Check if this is a 'LogMessage' response. Should we log these in verbose mode?
+      json::Value methodJson = responseJson["method"];
+      if (methodJson.isString())
+      {
+         std::string method = methodJson.getString();
+         if (method == "LogMessage")
+            continue;
+      }
+
+      // Check the response id. This will be missing for notifications; we may receive
+      // a flurry of progress notifications when requesting completions from Copilot.
+      // We might want to handle these somehow. Perhaps hook them up to some status widget?
+      json::Value requestIdJson = responseJson["id"];
+      if (!requestIdJson.isString())
+      {
+         DLOG("Ignoring response with missing id.");
+         continue;
+      }
+
+      // Check for a continuation handler, and invoke it if available.
+      std::string requestId = requestIdJson.getString();
+      if (s_pendingContinuations.count(requestId))
```

Comment:
Should we log a warning if we get a response without a known continuation handler?